src/mds/Server.cc

```diff
@@ -10938,6 +10940,7 @@ void Server::handle_client_mksnap(MDRequestRef& mdr)
     em.first->second = info;
   newsnap.seq = snapid;
   newsnap.last_created = snapid;
+  newsnap.last_modified = info.stamp;
```

Comment:
Since the `sr_t::last_modified` should be monotonically increasing, what if the `info.stamp` is smaller ? This is possibly happening when the request is deferred due to some reasons, etc.

Reply:
This makes it way to the client and ends up getting assigned to snapdir's mtime/ctime. What if we use the current (monotonic) timestamp here? That should be fine since snapdir's {a,m}time can't be updated.

Reply:
Yeah, the clients will ignore it anyway. But for a new mount it won't. Then in two different mount the `ctime` could be different for the same `snapdir`.

I saw many places in `mds/Server.cc` is doing like this:

```
11203   pi.inode->ctime = mdr->get_op_stamp();
11204   if (mdr->get_op_stamp() > pi.inode->rstat.rctime)                                                                                              
11205     pi.inode->rstat.rctime = mdr->get_op_stamp();
```

Reply:
What if `last_modified` isn't updated when its ahead of `mdr->get_op_stamp()`? Snapdir'd mtime/ctime won't change then.

Reply:
> Yeah, the clients will ignore it anyway. But for a new mount it won't. Then in two different mounts the `ctime` could be different for the same `snapdir`.

Right -- but, currently, {m,c}time is broken for snapdir anyway.

Reply:
> What if `last_modified` isn't updated when its ahead of `mdr->get_op_stamp()`? Snapdir'd mtime/ctime won't change then.

My understanding is that:

Case1:
1, if mksnap requestA is deferred.
2, a new mksnap requestB comes with a newer timestamp and updated the `last_modified`
3, the mksnap requestA continues and it won't update the timestamp.

In this case if these two requests from different clients, both of them will see the `ctime/mtime` updated, because the snapream will notify them twice.

Case2:
1, if  mksnap requestA is deferred.
2, a new mksnap requestB comes with a newer timestamp and is deferred too.
3, the mksnap requestA continues and it will update the `last_modified`
4, mksnap requestB continues and it will update the `last_modified` too.

And the client(s) will see the `mtime/ctime` updated too.

I am not sure whether the sequence in Case1 could happen in mksnap, but it seems could happen in other places as I remembered when fixing bugs before, I couldn't recall exactly which one.

And I am not sure why some code in `mds/Server.cc` is just doing:

`11203 pi.inode->ctime = mdr->get_op_stamp();`

So if your current code is an issue here, then many other places.

I am okay with your current code, just a little confused about the Case1 why we didn't handle that.

It's up to you for this. I am okay with it.

Reply:
> > What if `last_modified` isn't updated when its ahead of `mdr->get_op_stamp()`? Snapdir'd mtime/ctime won't change then.
> 
> My understanding is that:
> 
> Case1: 1, if mksnap requestA is deferred. 2, a new mksnap requestB comes with a newer timestamp and updated the `last_modified` 3, the mksnap requestA continues and it won't update the timestamp.
> 
> In this case if these two requests from different clients, both of them will see the `ctime/mtime` updated, because the snapream will notify them twice.

Right. But, NFS will cache snapdir's mtime (and change_attr) when it does a readdir on snapdir with requestA still deferred. When requestA is done snapdir's mtime/change_attr does not change (iff `last_modified` isn't updated when its ahead of requestA `mdr->get_op_stamp()`). If NFS serves directory entries from its cache (since mtime/change_attr is not updated), that would serve stale data.

> 
> Case2: 1, if mksnap requestA is deferred. 2, a new mksnap requestB comes with a newer timestamp and is deferred too. 3, the mksnap requestA continues and it will update the `last_modified` 4, mksnap requestB continues and it will update the `last_modified` too.
> 
> And the client(s) will see the `mtime/ctime` updated too.
> 
> I am not sure whether the sequence in Case1 could happen in mksnap, but it seems could happen in other places as I remembered when fixing bugs before, I couldn't recall exactly which one.

Hmmm.. Its all over the place in the MDS.

> 
> And I am not sure why some code in `mds/Server.cc` is just doing:
> 
> `11203 pi.inode->ctime = mdr->get_op_stamp();`
> 
> So if your current code is an issue here, then many other places.

If we want to fix this, then we should be fixing _all_ other instances.

> 
> I am okay with your current code, just a little confused about the Case1 why we didn't handle that.
> 
> It's up to you for this.

MDCache::predirty_journal_parents() uses `mdr->get_op_stamp()` for mtime, which is why I kept it this way.

Reply:
Yeah, so let's do that in future if this is really a bug.